evm/src/contracts/CircleIntegration/Governance.sol

```diff
@@ -0,0 +1,206 @@
+import {IWormhole} from "src/interfaces/IWormhole.sol";
+import {IGovernance} from "src/interfaces/IGovernance.sol";
+
+import {State} from "./State.sol";
+import {
+    SequentialRegistration,
+    getRegisteredEmitters,
+    getChainToDomain,
+    getConsumedVaas,
+    getDomainToChain,
+    getSequentialRegistration
+} from "./Storage.sol";
+
+abstract contract Governance is IGovernance, State, ERC1967Upgrade {
+    using BytesParsing for bytes;
+
+    uint16 constant GOVERNANCE_CHAIN = 1;
+    bytes32 constant GOVERNANCE_EMITTER =
+        0x0000000000000000000000000000000000000000000000000000000000000004;
+
+    // "CircleIntegration" (left-zero-padded)
+    bytes32 constant GOVERNANCE_MODULE =
+        0x000000000000000000000000000000436972636c65496e746567726174696f6e;
+
+    uint8 constant REGISTER_EMITTER_AND_DOMAIN = 2;
+    uint8 constant CONTRACT_UPGRADE = 3;
+
+    /// @inheritdoc IGovernance
+    function registerEmitterAndDomain(bytes memory encodedVaa) public {
+        (IWormhole.VM memory vaa, uint256 offset) =
+            _verifyAndConsumeGovernanceMessage(encodedVaa, REGISTER_EMITTER_AND_DOMAIN);
+
+        // Registering emitters should only be relevant for this contract's chain ID,
+        // unless the target chain is 0 (which means all chains).
+        uint16 targetChain;
+        (targetChain, offset) = vaa.payload.asUint16Unchecked(offset);
+        require(targetChain == 0 || targetChain == _chainId, "invalid target chain");
+
+        uint16 foreignChain;
+        (foreignChain, offset) = vaa.payload.asUint16Unchecked(offset);
+        require(foreignChain != 0 && foreignChain != _chainId, "invalid chain");
+
+        mapping(uint16 => bytes32) storage registeredEmitters = getRegisteredEmitters();
+
+        // For now, ensure that we cannot register the same foreign chain again.
+        require(registeredEmitters[foreignChain] == 0, "chain already registered");
+
+        bytes32 foreignAddress;
+        (foreignAddress, offset) = vaa.payload.asBytes32Unchecked(offset);
+        require(foreignAddress != 0, "emitter cannot be zero address");
+
+        uint32 cctpDomain;
+        (cctpDomain, offset) = vaa.payload.asUint32Unchecked(offset);
+        require(cctpDomain != _localCctpDomain, "domain == localDomain()");
+
+        _checkLength(vaa.payload, offset);
+
+        // Set the registeredEmitters state variable.
+        registeredEmitters[foreignChain] = foreignAddress;
+
+        // Ideally we would like to be able to update foreign emitters with governance VAAs instead
+        // of having to upgrade the contract to change this contract's state. But currently
+        // governance VAA sequence numbers and timestamps are not reliable enough to order these
+        // registrations. Until either value becomes reliable, we will store the timestamp in the
+        // sequential registration storage slot.
+        mapping(uint16 => SequentialRegistration) storage sequentialRegistration =
```

Comment:
I don't understand the purpose of this sequential registration logic? Why would we ever have 2 or more foreign emitter updates for the same remote chain happening at the same time where ordering could be a problem?

Since we're consuming these governance VAA hashes is it not safe to allow calls to this method even if the foreign chain has already been registered (i.e. to update the registered emitter)?

Reply:
The idea here is to potentially allow sequential updates for registered emitters in the future if there were a reliable way to sequence these governance VAAs. Right now sequence numbers are randomized and timestamp is unreliable (all of these sorts of registrations, including Token Bridge emitters, have timestamp 0 encoded in the VAA).

We may never go down this route and just upgrade the contract to update an emitter. I have no strong opinions about what we do. I can remove this entirely and revisit later. What do you think?

Reply:
Removed it here: https://github.com/wormhole-foundation/wormhole-circle-integration/pull/48/commits/f32f514388934f17515db85d4e31a32c4b500db3

Reply:
Having or not having this functionality now won't impact the ability to allow sequential updates in the future right? So we'd still need to do another upgrade if/when we're able to support this?

If so I'd suggest getting rid of this for now...firstly it would save gas and secondly I personally don't like having unused logic around that we might have to change later anyway